ios/RCTMapboxGL/RCTMapboxGLManager.m

```diff
@@ -612,22 +612,76 @@ - (NSArray*)serializePacksArray:(NSArray<MGLOfflinePack*>*)packs
     [_bridge.uiManager addUIBlock:^(RCTUIManager *uiManager, NSDictionary<NSNumber *, RCTMapboxGL *> *viewRegistry) {
         RCTMapboxGL *mapView = viewRegistry[reactTag];
         if ([mapView isKindOfClass:[RCTMapboxGL class]]) {
-            NSNumber *screenCoordX = options[@"screenCoordX"];
-            NSNumber *screenCoordY = options[@"screenCoordY"];
-            if (!screenCoordX || !screenCoordY) {
+            NSDictionary *pointDict = options[@"point"];
+            NSDictionary *rect = options[@"rect"];
+            if (!pointDict) {
                 // TODO: figure out best way to error this. Should we be using RCTPromise rather than a callback?
 //                reject(@"invalid_arguments", @"queryRenderedFeatures(): screenCoordX and screenCoordY are required.", nil);
                 return;
             }
+            NSNumber *screenCoordX = pointDict[@"screenCoordX"];
+            NSNumber *screenCoordY = pointDict[@"screenCoordY"];
+            NSSet<NSString *> *styleLayerIdentifiers = options[@"layers"];
+
+            CLLocationCoordinate2D location = [_map convertPoint:[sender locationInView:_map] toCoordinateFromView:_map];
+            CGPoint screenCoord = [sender locationInView:_map];
+
+
             CGPoint point = CGPointMake(screenCoordX.floatValue, screenCoordY.floatValue);
 
-            NSArray * features = [mapView visibleFeaturesAtPoint:point];
+            NSArray *features;
+            if (!styleLayerIdentifiers) {
+                features = [mapView visibleFeaturesAtPoint:point];
+            } else {
+                features = [mapView visibleFeaturesAtPoint:point inStyleLayersWithIdentifiers:styleLayerIdentifiers];
+            }
+
+            NSDictionary *geoJSONTypesByMGLClassName = @{
+                    @"MGLPointFeature": @"Point",
+                    @"MGLPolylineFeature": @"LineString",
+                    @"MGLPolygonFeature": @"Polygon"
+            };
 
             // TODO: also accept Rect to use [mapView visibleFeaturesInRect:]
 
-            // TODO: convert features to JS friendly objects
+            // TODO: switch to using MGLFeature's geoJSONFeature method if implemented - https://github.com/mapbox/mapbox-gl-native/issues/6302
+            NSMutableArray *geoJSONFeatures = [[NSMutableArray alloc] init];
```

Comment:
Nit: you can also write this more succinctly as `[NSMutableArray array]`. As a possibly premature but reasonable optimization, you can also specify the size upfront: `[NSMutableArray arrayWithCapacity:features.count]`.